Inital GCP object storage support

elcallio opened the pull request:
Adds infrastructure and client for interaction with GCP object storage services. 

Note: this is just a client object usable for creating, listing, deleting and up/downloading of objects to/from said storage service. It makes no attempt at actually inserting it into the sstable storage flow. That can come later. 

This PR breaks out GCP auth and some general REST call functionality into shared routines. Not all code is 100% reused, but at least some. 

Test is added, though could be more comprehensive (feel free to suggest a test vector).
Test can run in either local mock server mode (default), or against actual GCP. 
See `test/boost/gcp_object_storage_test.cc` for explanation on the config environment vars. 
Default is to run the test against a temporary docker deamon.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17926/) | :white_check_mark: | sif |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17926/) | :x: | i-0da70b5a44ffa7b3f |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17926/) | :x: | i-0ed7eb933fd831807 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17926/) | :x: | i-0fd06c40de2268fb9 |

#### Build Details:
- Duration: 29 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | ran |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/gcp_object_storage_test.cc<br> :small_blue_diamond: test/boost/test_config.yaml | :white_check_mark: | spider2.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider9.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/gcp_object_storage_test.cc<br> :small_blue_diamond: test/boost/test_config.yaml | :white_check_mark: | spider9.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17929/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/gcp_object_storage_test.cc<br> :small_blue_diamond: test/boost/test_config.yaml | :white_check_mark: | spider8.cloudius-systems.com |
| dtest | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17000/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8914/) | :white_check_mark: |  |
|  | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5725/) | :white_check_mark: |  |

#### Build Details:
- Duration: 1 hr 57 min

xemul commented:
Some "architectural" note is welcome. In particular -- the newly introduced code uses seastar http client, but it as well needs utils/rest/ functionality which AFAIK instantiates temporary client object to make requests. Why does gcp client need both?

xemul commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+            throw;
+        } catch (...) {
+            // network, whatnot. maybe add retries here as well, but should really 
+            // be on seastar level
```
Yes. https://github.com/scylladb/seastar/issues/2803

kreuzerkrieg replied:
They use the same approach as AWS does, what does http code 429 mean here? In fact it is 5 different erroneous states. Some are retryable, some not, in any case, you will be unable to determine the root cause of the failure once one http code has multiple meanings, so how are you going to handle it in response callbacks (I guess this is what you mean by "calling routines"), each time `switch...case` http codes and parsing erroneous responses?

elcallio replied:
We don't really have to handle error codes very specificly. Ok and created, and partial depending on context. 5xx should be retried (yes, should eventually move to shared code maybe). As for 429 etc; Not sure we need that level of specificity. Maybe adding extracting error messages, yes, but that is not retryability, just info for throwing.

kreuzerkrieg replied:
It is not about 429 or any other http code we can get from the server. My point is simple, once you get a 4XX code, you log it and then, in your opinion, the engineer investigating failure and searching through logs will understand what exactly went wrong on GSC side when encountering such a log line?
For example, Forbidded has 15 reasons to be sent to the client
https://cloud.google.com/storage/docs/xml-api/reference-status#403%E2%80%94forbidden

elcallio commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+ * Performs a REST post/put/get with credential refresh/retry.
+ */
+future<> 
+utils::gcp::storage::client::impl::send_with_retry(const std::string& path, const std::string& scope, body_variant body, std::string_view content_type, handler_func_ex handler, httpclient::method_type op, key_values headers) {
```
No, it is similar to it. Note the "aws" in aws::retryable_http_client. And all the AWS error checks etc in it. Not very usable here.

elcallio commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+        } catch (httpd::unexpected_status_error& e) {
+            gcp_storage.debug("{}: Got unexpected response: {}", _endpoint, e.status());
+            if (e.status() == status_type::unauthorized && retries++ < 3) {
+                // refresh access token and retry.
```
See the call at the head of the loop to `_credentials->refresh`. If we fail due to auth token becoming invalid during the call, loop back, refresh the token, and retry.

elcallio commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+        }
+    }
+    future<> put(temporary_buffer<char> buf) override {
+        _buffers.emplace_back(std::move(buf));
```
Possibly. Though, we really _do_ want to accumulate ~8MB or more, since that is the "optimal" upload size for a given chunk. And behaviour here would probably be dictated by the overall parallelism of writing; If we are highly concurrent (writing many objects in parallel), we'd might want to limit the buffering to something like 2*8MB, and block until previous chunk is done at that point. But if we are wholly singular, maybe we don't...

kreuzerkrieg replied:
why 8? why to limit to 2*8 if we can afford more? for example, for s3 client we do the following `std::clamp<size_t>(memory::stats().total_memory() * 0.01, 10 << 20, 100 << 20);`

kreuzerkrieg replied:
if it is a huge object, and this is the only object you work on in this shard why not to optimize your put request with huge buffers?
For example we found out that minimal upload part in S3 (5MiB) provides us with suboptimal throughput

elcallio replied:
But if the limit is huge, does it even make any difference? That is my point.

elcallio commented on utils/rest/client.cc:
```diff
@@ -208,6 +208,28 @@ future<> rest::send_request(std::string_view uri
 
+std::tuple<std::string, std::string, bool, uint16_t> rest::parse_simple_uri(std::string_view uri) {
+        // Extremely simplified URI parsing. Does not handle any params etc. But we do not expect such here.
+    static boost::regex simple_url(R"foo((https?):\/\/([^\/:]+)(:\d+)?(\/.*)?)foo");
```
That is a very good point. I sometimes forget we depend on the full set of boost packages... :-)

elcallio replied:
Boost regex I _do_ remember. :-) Anyhow, code becomes nicer adapting boost::url. Thanks.

elcallio replied:
Could not use boost, but moved and structified the helpers to make it a little more readable.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+
+    boost::smatch m;
+    std::string tmp(i->second);
+    if (!boost::regex_match(tmp, m, range_ex)) {
```
Why not `std::regex_match`?

elcallio replied:
std::regex has a number of issues (esp. in our stdlib iirc), so most scylla code uses boost::regex instead.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+                }
+                if (reply::classify_status(res.result()) == reply::status_class::server_error && ++retries < 4) {
+                    gcp_storage.debug("{}:{} chunk {}:{} server error. backoff {}...", _bucket, _object_name, offset, offset+len, retries);
+                    co_await backoff(retries);
```
They mention exponential backoffs in their documentation

elcallio replied:
Done.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+            break;
+        } catch (httpd::unexpected_status_error& e) {
+            gcp_storage.debug("{}: Got unexpected response: {}", _endpoint, e.status());
+            if (e.status() == status_type::unauthorized && retries++ < 3) {
```
`e.status() == status_type::unauthorized`
401 has 4 meanings and only 2 could possibly be solved with re-authenticating

kreuzerkrieg replied:
`retries++ < 3`
magic number

elcallio replied:
In the case of OAuth, it can only have one meaning here really. If the credentials are truly invalid, not just expired, the refresh we do before/after should fail.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+                    gcp_storage.debug("{}:{} chunk {}:{} done", _bucket, _object_name, offset, offset+len);
+                    co_return;
+                }
+                if (reply::classify_status(res.result()) == reply::status_class::server_error && ++retries < 4) {
```
What about all "4xx" retryables?

elcallio replied:
Added to backoff-retry.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+    }
+}
+
+// See https://cloud.google.com/storage/docs/listing-buckets
+future<std::vector<utils::gcp::storage::object_info>> utils::gcp::storage::client::list_objects(std::string_view bucket, std::string_view prefix) {
```
you have to align here, list what, buckets or objects?

elcallio replied:
Objects. As the method is named. The doc name means "list _contents_ of buckets", but that is just another name for objects.

kreuzerkrieg commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+            if (!items->IsArray()) {
+                throw failed_operation("Malformed list object items");
+            }
+            for (auto& item : items->GetArray()) {
```
Are we sure scylladb will no tgo haywire with a huge number of objects in the bucket? Maybe we want pagination here?

kreuzerkrieg replied:
I think it was @bhalevy mentioning a customer with 600k sstables, now think about the S3 backup folder for such a cluster

elcallio replied:
Flat bucket? Even with listing streaming that is a rather hefty operation...

elcallio commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+future<temporary_buffer<char>> utils::gcp::storage::client::object_data_source::get() {
+    // If we don't know the source size yet, get the info from server
+    if (_size == 0) {
+        co_await read_info();
```
Docs does not say anything about actual object size being returned by a read. I agree that one could just read until it return nothing, true. But the problem is that I also need to know the object generation number. Otherwise, we could start reading, someone replaces the object mid-read, and then we read other data for parts of the download.

elcallio replied:
Yes.

elcallio commented on utils/gcp/object_storage.cc:
```diff
@@ -0,0 +1,820 @@
+    return seastar::sleep(n * 300ms);
+}
+
+static const boost::regex range_ex("bytes=(\\d+)-(\\d+)");
```
Yes. But the returned range in this case will always follow the format here.

elcallio commented on test/boost/gcp_object_storage_test.cc:
```diff
@@ -0,0 +1,349 @@
+    });
+}
+
+static future<seastar::experimental::process> start_fake_gcs_server(const tmpdir& tmp) {
```
Mainly that we have better handling for json, as much as I don't love rapidjson, it is fairly integrated. For previous GCP interop we've used json, so momentum also.

elcallio replied:
See comment below, and https://docs.simplebackups.com/storage-providers/ghsg5GE15AMwMo1qFjUCXn/google-cloud-storage-s3-compatible-credentials/8ZKUSSJRJxA4mU4VdxvRfo. You will at least need to create extraneous credentials etc. The API is probably fine, but a client will not be able to use it "out-of-the-box" as such. 
I can also not imagine that googles "own" API:s are not more efficient/maintained than interop API:s for their own product.

elcallio commented on utils/rest/client.cc:
```diff
@@ -121,6 +124,90 @@ void rest::httpclient::target(std::string_view s) {
+
+    uint16_t pi = port.empty() ? (creds ? 443 : 80) : uint16_t(std::stoi(port.substr(1)));
+
+    httpclient client(host, pi, std::move(creds));
```
The use case for the rest::client are connections that are few and far between (mainly EAR and auth atm). Note that the GCP client does _not_ use it, since we assume it will have more frequent connections.

avikivity replied:
@xemul don't we have a pooling http client? Though maybe it doesn't allow sharing a pool among different endpoints.

elcallio commented:
Not sure what you mean... The fake-gcs-server is from https://github.com/fsouza/fake-gcs-server/. It is far from a perfect impl of GCP, but good enough for our testing purposes I'd say. Also, the tests can also run against "real" GCP if parameterized via env for auth etc, so verification of the functionality used is not limited to the mock. 
Typically, I'd assume we parameterize jenkins to run with real GCP (as tests erase their objects etc -> should not add much cost), at least some of the time.

kreuzerkrieg commented:
> Not sure what you mean... The fake-gcs-server is from https://github.com/fsouza/fake-gcs-server/. It is far from a perfect impl of GCP, but good enough for our testing purposes I'd say. Also, the tests can also run against "real" GCP if parameterized via env for auth etc, so verification of the functionality used is not limited to the mock. Typically, I'd assume we parameterize jenkins to run with real GCP (as tests erase their objects etc -> should not add much cost), at least some of the time.

ah, I overlooked this one, then all good

mykaul commented:
Nice to see the build failure right here in a comment!

fruch commented:
> Ping @scylladb/qa-team ... The CI above failed because the jenkins running apparently could not find a docker installed/available for tests. This seems odd, given that at least dtests (and unit tests?) use docker quite a bit... Am I missing something?

your tests are running within dbuild, I don't think anyone is installing docker inside of it.
maybe podman would work ? but I would recommend testing it locally with dbuild 

dtests are running on their own environment which does support docker/podman.

mykaul commented:
> > Ping @scylladb/qa-team ... The CI above failed because the jenkins running apparently could not find a docker installed/available for tests. This seems odd, given that at least dtests (and unit tests?) use docker quite a bit... Am I missing something?
> 
> your tests are running within dbuild, I don't think anyone is installing docker inside of it. maybe podman would work ? but I would recommend testing it locally with dbuild
> 
> dtests are running on their own environment which does support docker/podman.

I thought we enabled nesting not so long ago.

avikivity commented:
> Afaik, even if using S3 compat, you will need to do extra hoops for authentication, i.e. set up "extra" authentication in your GCP project to provide AWS-compatible credentials. The idea here (I assume) was to provide a direct GCP API, using google oauth etc, as well as the direct GCP object storage API:s.

I'm not against it. Main concern is local testing, we have minio for S3, what do we have for GCP? Will go look.

avikivity commented on test/boost/gcp_object_storage_test.cc:
```diff
@@ -0,0 +1,334 @@
+        exec = tp::find_file_in_path("docker");
+        if (exec.empty()) {
+            exec = tp::find_file_in_path("podman");
+            image = "docker.io/fsouza/fake-gcs-server";
```
Better to always use the fully qualified image. Both with the repository, and with the label (:something), so it's effectively frozen.

I see you are brave enough to be the first to try nested container support, let's hope it works.

avikivity replied:
And I see they kindly provide aaaaarch64 versions too.

avikivity commented on test/boost/gcp_object_storage_test.cc:
```diff
@@ -0,0 +1,334 @@
+        }
+        params = {
+            "",
+            "run", "--rm", "--name", "fake-gcs-server", "-p", "4443:4443",
```
Need to allocate either a random bind address or a random port so parallel runs can work.

avikivity replied:
The server is immaterial here since you bind whatever port in the container to a port in the host (which is also a container under dbuild).

avikivity commented on test/boost/test_config.yaml:
```diff
@@ -16,6 +16,7 @@ run_first:
 no_parallel_cases:
     - logalloc_test
     - logalloc_standard_allocator_segment_pool_backend_test
+    - gcp_object_storage_test
```
If the fixtuxuture would be external to the process, this wouldn't be necessary (but not so important).

avikivity commented:
> The GCP object storage emulator used here. "fake-gcs-server". Which however I am right now having issues with due to docker (suddenly?) not being available,

We don't have (and never had) docker in the frozen toolchain.

> and podman having fuse-overlay issues (?)

I did some magic to get podman to run in a nested container (exactly for this use case), but the magic only works with the right phase of the moon. See 2d2a2ef277725b06ca2dd10ff6abb87be2edfdbf, 4c1f4c419c67136498f687f809f9fbc818fb0276.

avikivity commented:
For example, this works for me:

```console
$ ./tools/toolchain/dbuild podman run --rm docker.io/fedora:42 whoami
root
```

elcallio commented:
Yes, I saw these. However, I still get podman complaining about fuse-overlay not being installed. Installing it (in an interactive image) makes the nested container run work.

elcallio commented:
> Can you give me a simple reproducer? If installing fuse-overlay is all that's needed, that's a simple fix.

Running on a fedora42:
```
$ ./tools/toolchain/dbuild  podman run --rm docker.io/fedora:42 whoami
Error: configure storage: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
```
```
$ ./tools/toolchain/dbuild -i --
bash-5.2# dnf install -y fuse-overlayfs
....
Trying to pull docker.io/library/fedora:42...
Getting image source signatures
Copying blob 4c5aa609a918 done   | 
Copying config a17b50ce31 done   | 
Writing manifest to image destination
root
bash-5.2#
```

I have no idea why it works differently for me. Kernel version? (6.14)...

Worth noting is also that trying any of this in a nested docker container (running dbuild with docker) fails in other wonderful ways. Which might again be too old a docker of course...

avikivity commented:
I reproduced the problem and the fix, but it's suboptimal - on my system the podman image cache is shared between the host and the container, so repeated invocations with the same image don't download it again.

avikivity commented:
I found that if I do this:

```
mv .local/share/containers/storage/libpod/db.sql{,.backup}
```

and certain other things which I still have to sift through, it works reliably and shares the image cache.

Perhaps I can bind-mount this directory to nothing and things will work, still need to try things out.

avikivity commented:
My own installation doesn't have this file.

elcallio commented:
The patch works for me on the clean fedora42. Even with "/home/avi" in the container patch. :-)

avikivity commented:
> The patch works for me on the clean fedora42. Even with "/home/avi" in the container patch. :-)

Maybe it's a clue that it can be pruned out.

avikivity commented:
https://github.com/scylladb/scylladb/pull/25156

elcallio commented:
Only rebased on latest master, partially to get docker-in-docker CI testing working again.

elcallio commented:
> The cover letter mentions lack of test coverage.

Not lack as such, only that I am not sure I exactly cover all possible cases (obviously not). There are only so many combinations of reading, writing and deleting objects that one can do before it is redundant. 
Once we start using it in actual sstable storage, there might pop up cases to add to unit tests, but atm I am not seeing them. Maybe add test for broken auth/bad permissions, but these only work on real GCP, which the CI does not yet handle, and the test itself would just be verifying that we _can't_ read/write a file...

elcallio commented:
@avikivity - debug test run failed again on missing overlayfs. 
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18768/artifact/testlog/x86_64/debug/gcp_object_storage_test_stdout.1.log

Did this execute on a machine that is even more differently abled? (All dbuild changes merged/rebased to afaik).

avikivity commented:
> I am more wondersome about the apparently somewhat heterogeneous environment CI is running in. It would be nice if things either worked on CI machines, or not...

I know they're planning to standardize on Fedora 42. But eventually they're move to 43 and 44 and I want the old toolchains to keep working.

avikivity commented:
> > I am more wondersome about the apparently somewhat heterogeneous environment CI is running in. It would be nice if things either worked on CI machines, or not...
> 
> What would be the fun in that?
> 
> @avikivity - wrt disabling cache - I think it's a taping over the issue of wrongly/partially cleaning the local image repo.

No, podman changes the metadata layout from time to time, and if we don't bind-mount everything exactly right, something is missing and it complains.

avikivity commented:
I tested dbuild on top of Fedora 41, and it worked.

avikivity commented:
> I am more wondersome about the apparently somewhat heterogeneous environment CI is running in. It would be nice if things either worked on CI machines, or not...

That was the whole purpose of dbuild, to isolate the host from CI. But nested podman is breaking that.

avikivity commented:
nested podman even runs on a Fedora 40 host.

elcallio commented:
Yes, it runs fine for me as well. But apparently not the CI host above...

fruch commented:
> @yaronkaikov is there a way to know which host the debug build ran on?

It say `spider4.cloudius-systems.com` right next to it.

in dtest we were using docker exactly like that, but we kept the host controlled and identical (so far), and the docker features we were using we relatively old enough, to not running into hosts that doesn't have them.

but dbuild was always depended on the host to have podman or docker, now it would be podman from specific version, isn't it ?

avikivity commented:
> > @yaronkaikov is there a way to know which host the debug build ran on?
> 
> It say `spider4.cloudius-systems.com` right next to it.

Right, I was looking at the console log.

> 
> in dtest we were using docker exactly like that, but we kept the host controlled and identical (so far), and the docker features we were using we relatively old enough, to not running into hosts that doesn't have them.

I think you're using a different method, forwarding the control socket and so the containers are not nested, but instead running on the host alongside the main container.


> 
> but dbuild was always depended on the host to have podman or docker, now it would be podman from specific version, isn't it ?

I'd really like not to depend on specific versions.

avikivity commented:
@yaronkaikov how can I gain access to spider4?

elcallio commented:
Pushed new version:

* Added inheriting process environment to spawned process. Not doing so can cause weirdness with path etc. 
* Added log check-wait for spawned fake-gcs-server, since when using podman (at least) container, the 4443 port will in fact be connectible even before the actual server is up, potentially causing spurious connection fails etc in test.

elcallio commented:
There is a limit to docker pull? And we hit it? (CI machine above)
```
Trying to pull docker.io/fsouza/fake-gcs-server:latest...
time="2025-08-05T15:25:34+03:00" level=warning msg="Failed, retrying in 1s ... (1/3). Error: initializing source docker://fsouza/fake-gcs-server:latest: reading manifest latest in docker.io/fsouza/fake-gcs-server: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit"
```

yaronkaikov commented:
> There is a limit to docker pull? And we hit it? (CI machine above)
> 
> ```
> Trying to pull docker.io/fsouza/fake-gcs-server:latest...
> time="2025-08-05T15:25:34+03:00" level=warning msg="Failed, retrying in 1s ... (1/3). Error: initializing source docker://fsouza/fake-gcs-server:latest: reading manifest latest in docker.io/fsouza/fake-gcs-server: toomanyrequests: You have reached your unauthenticated pull rate limit. https://www.docker.com/increase-rate-limit"
> ```

@elcallio is the test using docker or podman to pull the images?

avikivity commented:
https://github.com/scylladb/scylladb/pull/25354

elcallio commented:
Rebased.

avikivity commented:
Nice to have two greens in a row, but can I have an explanation of the problems and fixes along the way?

I want to be sure this isn't a flaky test so I don't doom next to a series of failures.

elcallio commented:
Three problems:

1.) Port/network. The fake-gcs-server image cannot pick a random port. Tried using serial test execution to prevent conflicting containers, but this breaks if the same machine runs debug/release in parallel etc. 
Then tried letting the podman/docker pick a forwarded port. Works, but not in dbuild due to using host network. 
Final solution is to pick a port using listen in the test, then try starting the container with this and hope it can bind. Rinse, repeat. Somewhat ugly, but works. 

2.) Image download. The timeout for container startup was to low for some runs, where getting the image took a few more seconds than expected. Bumped this timeout quite a bit. 

3.) A lame string match typo that caused the above timeout to be ignored anyway, because we triggered "ready" to early. With the correct match, we wait properly for image + container start.

elcallio commented:
ping?

avikivity commented on test/boost/gcp_object_storage_test.cc:
```diff
@@ -0,0 +1,456 @@
+        params = {
+            "",
+            "run", "--rm", "-p", "--", // set below
+            "docker.io/fsouza/fake-gcs-server"
```
Please use a tag here (e.g. :1.52.3) so we aren't exposed to random regressions.

elcallio replied:
Done.

elcallio commented:
ping?

elcallio commented:
Nameful pings are above. Just re-pinging...

avikivity commented:
> ping?

What changed?

elcallio commented:
As stated above (and in commit message); added explicit image version.

avikivity commented:
Let's hope the nested container thing doesn't blow up in the future.

yaronkaikov commented:
> #### 🟢 CI State: SUCCESS
> Mode	Stage	Status	Node
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider9.cloudius-systems.com
> dev	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider6.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider6.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)
> 🔹 test/boost/gcp_object_storage_test.cc	✅	spider6.cloudius-systems.com
> release	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider12.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider12.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)
> 🔹 test/boost/gcp_object_storage_test.cc	✅	spider12.cloudius-systems.com
> debug	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider8.cloudius-systems.com
> [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)	✅	spider8.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/19425/)
> 🔹 test/boost/gcp_object_storage_test.cc	✅	spider8.cloudius-systems.com
> dtest	[dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/10194/)	✅	
> [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/6949/)	✅	
> [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/18303/)	✅	
> #### Build Details:
> * Duration: 7 hr 34 min

This CI run failed, we have some critical bug in the process (which we are working on), i am not sure if those failure related to this PR or not, better to check and or dequeue for now :-(

elcallio commented:
> This CI run failed, we have some critical bug in the process (which we are working on), i am not sure if those failure related to this PR or not, better to check and or dequeue for now :-(

Not to complain or anything, but why are the markers green? 
Anyway, it looks like the failures are in vector store tests. Seems unrelated. Though I guess it would still be prudent to await a fix to rebase on.

elcallio commented:
Rebased and re-triggered.

avikivity commented:
Well, it's already queued.
